File: backend/app/services/inventory_service.py
"""
Servicio de lógica de negocio para gestión de inventario.
"""
from typing import List, Optional
from datetime import datetime
from sqlalchemy.orm import Session
from sqlalchemy import func
from fastapi import HTTPException, status

from app.models.inventory_movement import InventoryMovement, MovementType, MovementReason
from app.models.product import Product
from app.repositories.inventory_repository import InventoryMovementRepository
from app.repositories.product_repository import ProductRepository
from app.schemas.inventory import (
    InventoryMovementCreate,
    InventoryMovementResponse,
    InventoryMovementList,
    InventoryMovementFilter,
    StockAdjustment,
    BatchStockEntry,
    BatchStockEntryRequest,
    LowStockProduct,
    LowStockAlert,
    InventoryStats,
    MovementTypeEnum,
    MovementReasonEnum,
)


class InventoryService:
    """Servicio para gestión de inventario."""

    def __init__(self, db: Session):
        self.db = db
        self.movement_repo = InventoryMovementRepository(db)
        self.product_repo = ProductRepository(db)

    def get_movement(self, movement_id: int) -> InventoryMovement:
        """Obtener un movimiento por ID."""
        movement = self.movement_repo.get_by_id(movement_id)
        if not movement:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Movimiento no encontrado"
            )
        return movement

    def get_movements(
        self,
        page: int = 1,
        page_size: int = 20,
        filters: Optional[InventoryMovementFilter] = None
    ) -> InventoryMovementList:
        """Obtener movimientos con paginación y filtros."""
        skip = (page - 1) * page_size
        movements, total = self.movement_repo.get_all(skip, page_size, filters)
        
        pages = (total + page_size - 1) // page_size if page_size > 0 else 0
        
        return InventoryMovementList(
            items=[InventoryMovementResponse.model_validate(m) for m in movements],
            total=total,
            page=page,
            page_size=page_size,
            pages=pages
        )

    def get_product_movements(
        self,
        product_id: int,
        limit: int = 50
    ) -> List[InventoryMovementResponse]:
        """Obtener historial de movimientos de un producto."""
        # Verificar que el producto existe
        product = self.product_repo.get_by_id(product_id)
        if not product:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Producto no encontrado"
            )
        
        movements = self.movement_repo.get_by_product(product_id, limit)
        return [InventoryMovementResponse.model_validate(m) for m in movements]

    def create_movement(
        self,
        data: InventoryMovementCreate,
        user_id: Optional[int] = None
    ) -> InventoryMovementResponse:
        """
        Crear un movimiento de inventario.
        Actualiza automáticamente el stock del producto.
        """
        # Obtener producto
        product = self.product_repo.get_by_id(data.product_id)
        if not product:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Producto no encontrado"
            )

        # Calcular nuevo stock
        stock_before = product.stock_current
        
        # Convertir enum de schema a enum de modelo
        movement_type = MovementType(data.movement_type.value)
        reason = MovementReason(data.reason.value)
        
        if movement_type == MovementType.ENTRY:
            stock_after = stock_before + data.quantity
        elif movement_type in (MovementType.EXIT, MovementType.ADJUSTMENT):
            stock_after = stock_before - data.quantity
            if stock_after < 0:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Stock insuficiente. Stock actual: {stock_before}, cantidad solicitada: {data.quantity}"
                )
        else:
            # Para transferencias u otros tipos futuros
            stock_after = stock_before

        # Crear movimiento
        movement = self.movement_repo.create(
            product_id=data.product_id,
            movement_type=movement_type,
            reason=reason,
            quantity=data.quantity,
            stock_before=stock_before,
            stock_after=stock_after,
            user_id=user_id,
            reference=data.reference,
            notes=data.notes
        )

        # Actualizar stock del producto
        self.product_repo.update_stock(product.id, stock_after)

        # Recargar con relaciones
        movement = self.movement_repo.get_by_id(movement.id)
        return InventoryMovementResponse.model_validate(movement)

    def add_stock(
        self,
        product_id: int,
        quantity: int,
        reason: MovementReasonEnum = MovementReasonEnum.PURCHASE,
        reference: Optional[str] = None,
        notes: Optional[str] = None,
        user_id: Optional[int] = None
    ) -> InventoryMovementResponse:
        """Agregar stock a un producto (entrada)."""
        data = InventoryMovementCreate(
            product_id=product_id,
            movement_type=MovementTypeEnum.ENTRY,
            reason=reason,
            quantity=quantity,
            reference=reference,
            notes=notes
        )
        return self.create_movement(data, user_id)

    def remove_stock(
        self,
        product_id: int,
        quantity: int,
        reason: MovementReasonEnum = MovementReasonEnum.SALE,
        reference: Optional[str] = None,
        notes: Optional[str] = None,
        user_id: Optional[int] = None
    ) -> InventoryMovementResponse:
        """Remover stock de un producto (salida)."""
        data = InventoryMovementCreate(
            product_id=product_id,
            movement_type=MovementTypeEnum.EXIT,
            reason=reason,
            quantity=quantity,
            reference=reference,
            notes=notes
        )
        return self.create_movement(data, user_id)

    def adjust_stock(
        self,
        data: StockAdjustment,
        user_id: Optional[int] = None
    ) -> InventoryMovementResponse:
        """
        Ajustar el stock de un producto a un valor específico.
        Crea un movimiento de ajuste automáticamente.
        """
        product = self.product_repo.get_by_id(data.product_id)
        if not product:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Producto no encontrado"
            )

        stock_before = product.stock_current
        stock_after = data.new_stock
        difference = abs(stock_after - stock_before)

        if difference == 0:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="El nuevo stock es igual al actual"
            )

        # Determinar tipo de movimiento
        if stock_after > stock_before:
            movement_type = MovementType.ENTRY
        else:
            movement_type = MovementType.ADJUSTMENT

        reason = MovementReason(data.reason.value)

        # Crear movimiento de ajuste
        movement = self.movement_repo.create(
            product_id=data.product_id,
            movement_type=movement_type,
            reason=reason,
            quantity=difference,
            stock_before=stock_before,
            stock_after=stock_after,
            user_id=user_id,
            reference=None,
            notes=data.notes or f"Ajuste de stock: {stock_before} → {stock_after}"
        )

        # Actualizar stock del producto
        self.product_repo.update_stock(product.id, stock_after)

        movement = self.movement_repo.get_by_id(movement.id)
        return InventoryMovementResponse.model_validate(movement)

    def batch_stock_entry(
        self,
        data: BatchStockEntryRequest,
        user_id: Optional[int] = None
    ) -> List[InventoryMovementResponse]:
        """
        Entrada masiva de stock (para compras).
        Procesa múltiples productos en una sola transacción.
        """
        results = []
        
        for item in data.items:
            movement = self.add_stock(
                product_id=item.product_id,
                quantity=item.quantity,
                reason=MovementReasonEnum.PURCHASE,
                reference=item.reference or data.reference,
                notes=item.notes,
                user_id=user_id
            )
            results.append(movement)
        
        return results

    def get_low_stock_products(self) -> LowStockAlert:
        """Obtener productos con bajo stock o sin stock."""
        # Query para productos con stock menor al mínimo
        products = (
            self.db.query(Product)
            .filter(Product.is_active == True)
            .filter(Product.stock_current < Product.stock_min)
            .order_by(
                (Product.stock_current - Product.stock_min).asc()  # Más críticos primero
            )
            .all()
        )

        low_stock_products = []
        critical_count = 0
        warning_count = 0

        for product in products:
            deficit = product.stock_min - product.stock_current
            
            if product.stock_current == 0:
                critical_count += 1
            else:
                warning_count += 1

            low_stock_products.append(LowStockProduct(
                id=product.id,
                sku=product.sku,
                name=product.name,
                stock_current=product.stock_current,
                stock_min=product.stock_min,
                stock_deficit=deficit,
                category_name=product.category.name if product.category else None,
                supplier_name=product.supplier.name if product.supplier else None
            ))

        return LowStockAlert(
            total_products=len(products),
            critical_count=critical_count,
            warning_count=warning_count,
            products=low_stock_products
        )

    def get_inventory_stats(self) -> InventoryStats:
        """Obtener estadísticas generales del inventario."""
        # Total de productos activos
        total_products = (
            self.db.query(func.count(Product.id))
            .filter(Product.is_active == True)
            .scalar() or 0
        )

        # Valor total del inventario (stock_current * cost)
        total_value = (
            self.db.query(func.sum(Product.stock_current * Product.cost))
            .filter(Product.is_active == True)
            .scalar() or 0
        )

        # Productos con bajo stock
        low_stock_count = (
            self.db.query(func.count(Product.id))
            .filter(Product.is_active == True)
            .filter(Product.stock_current < Product.stock_min)
            .filter(Product.stock_current > 0)
            .scalar() or 0
        )

        # Productos sin stock
        out_of_stock_count = (
            self.db.query(func.count(Product.id))
            .filter(Product.is_active == True)
            .filter(Product.stock_current == 0)
            .scalar() or 0
        )

        # Movimientos por período
        movements_today = self.movement_repo.count_today()
        movements_week = self.movement_repo.count_this_week()
        movements_month = self.movement_repo.count_this_month()

        return InventoryStats(
            total_products=total_products,
            total_stock_value=float(total_value),
            low_stock_count=low_stock_count,
            out_of_stock_count=out_of_stock_count,
            movements_today=movements_today,
            movements_this_week=movements_week,
            movements_this_month=movements_month
        )

    def validate_stock_available(self, product_id: int, quantity: int) -> bool:
        """Validar si hay stock disponible para una cantidad específica."""
        product = self.product_repo.get_by_id(product_id)
        if not product:
            return False
        return product.stock_current >= quantity

File: frontend/src/features/inventory/services/inventoryService.ts
/**
 * Servicio para gestión de inventario
 */
import apiClient from '@/shared/services/api'
import {
  InventoryMovementWithProduct,
  MovementListResponse,
  InventoryMovementCreate,
  StockAdjustment,
  BatchStockEntryRequest,
  LowStockAlert,
  InventoryStats,
  MovementFilter
} from '../types'

class InventoryService {
  private readonly basePath = '/api/v1/inventory'

  /**
   * Obtener movimientos con paginación y filtros
   */
  async getMovements(
    page: number = 1,
    pageSize: number = 20,
    filters?: MovementFilter
  ): Promise<MovementListResponse> {
    const params: Record<string, unknown> = {
      page,
      page_size: pageSize,
    }

    if (filters) {
      if (filters.product_id) params.product_id = filters.product_id
      if (filters.movement_type) params.movement_type = filters.movement_type
      if (filters.reason) params.reason = filters.reason
      if (filters.user_id) params.user_id = filters.user_id
      if (filters.reference) params.reference = filters.reference
      if (filters.date_from) params.date_from = filters.date_from
      if (filters.date_to) params.date_to = filters.date_to
    }

    const response = await apiClient.get<MovementListResponse>(`${this.basePath}/movements`, { params })
    return response.data
  }

  /**
   * Obtener un movimiento por ID
   */
  async getMovement(id: number): Promise<InventoryMovementWithProduct> {
    const response = await apiClient.get<InventoryMovementWithProduct>(`${this.basePath}/movements/${id}`)
    return response.data
  }

  /**
   * Crear un nuevo movimiento de inventario
   */
  async createMovement(data: InventoryMovementCreate): Promise<InventoryMovementWithProduct> {
    const response = await apiClient.post<InventoryMovementWithProduct>(`${this.basePath}/movements`, data)
    return response.data
  }

  /**
   * Obtener historial de movimientos de un producto
   */
  async getProductMovements(productId: number, limit: number = 50): Promise<InventoryMovementWithProduct[]> {
    const response = await apiClient.get<InventoryMovementWithProduct[]>(
      `${this.basePath}/products/${productId}/movements`,
      { params: { limit } }
    )
    return response.data
  }

  /**
   * Ajustar stock de un producto
   */
  async adjustStock(data: StockAdjustment): Promise<InventoryMovementWithProduct> {
    const response = await apiClient.post<InventoryMovementWithProduct>(`${this.basePath}/adjust`, data)
    return response.data
  }

  /**
   * Entrada masiva de stock (compras)
   */
  async batchStockEntry(data: BatchStockEntryRequest): Promise<InventoryMovementWithProduct[]> {
    const response = await apiClient.post<InventoryMovementWithProduct[]>(`${this.basePath}/batch-entry`, data)
    return response.data
  }

  /**
   * Obtener alertas de bajo stock
   */
  async getLowStockAlerts(): Promise<LowStockAlert> {
    const response = await apiClient.get<LowStockAlert>(`${this.basePath}/alerts/low-stock`)
    return response.data
  }

  /**
   * Obtener estadísticas de inventario
   */
  async getStats(): Promise<InventoryStats> {
    const response = await apiClient.get<InventoryStats>(`${this.basePath}/stats`)
    return response.data
  }

  /**
   * Verificar stock disponible
   */
  async checkStock(productId: number, quantity: number): Promise<{ available: boolean }> {
    const response = await apiClient.get<{ product_id: number; quantity_requested: number; available: boolean }>(
      `${this.basePath}/check-stock/${productId}`,
      { params: { quantity } }
    )
    return { available: response.data.available }
  }
}

export const inventoryService = new InventoryService()
export default inventoryService
